### PR TITLE
BLT Private Project 

### PR DESCRIPTION
**Pull request by HanilJain**
### description
BLT private project have Private Issue Page where all the Private Issue are listed and viewership can be managed.
request made by any user and people who have access are listed which can be granted or declined the access.
User could request access for Private Issue with message and the owner will notified by the mail , If access is granted the user will be notified through email and can view the Private Issue.

### Feature

- Request Access for Private Issue
- Email Notification on Request Request and approval
- viewership management of Private Issue 
- dedicated page for Private Issue management
- Well Tested and security management

### workflow
Screenshot 2024-02-04 050207

### Screnshot
Request Access

Private Issue on User Profile

Private Issue Listed

Viewership Managment

### CODE REVIEW

**Comment by HanilJain** [comment]
@DonnieBLT Sir , could be please review

**Comment by HanilJain** [comment]
This has both #1691 #1729 with additional feature fixing #1090 #1685.

**Comment by DonnieBLT** [comment]
I think this may need an update / conflict check

**Comment by DonnieBLT** [comment]
Can you please resolve the conflicts and update the branch?

**Comment by HanilJain** [comment]
@DonnieBLT Done please check.

**Comment by HanilJain** [comment]
@DonnieBLT are there any more changes to be made in it ?

**Comment by AtmegaBuzz** [comment]
Tested it and it works perfect. 
Some Fixes:
1. Use tailwind instead of CSS and bootstrap in added html (we are migrating to tailwind)
2. /issue/pk page is deprecated and will be removed in future please also add support for  /issue2/pk
3. change model name RequestAccess -> RequestIssueAccess (in future we will add access models for company and domains also, better we keep naming convention good)

**Review by DonnieBLT — CHANGES_REQUESTED** [review]
Please see comments for changes

**Comment by HanilJain** [comment]
> Tested it and it works perfect. Some Fixes:
> 
> 1. Use tailwind instead of CSS and bootstrap in added html (we are migrating to tailwind)
> 2. /issue/pk page is deprecated and will be removed in future please also add support for  /issue2/pk
> 3. change model name RequestAccess -> RequestIssueAccess (in future we will add access models for company and domains also, better we keep naming convention good)

@AtmegaBuzz changes have been made in 79c67d6fcdffb90bb6a7d025c949a5ef1512070d 
could you Please review

**Comment by AtmegaBuzz** [comment]
@HanilJain can you resolve the merge conflicts, rest looks good to me

**Review by DonnieBLT — CHANGES_REQUESTED** [review]

**Review by DonnieBLT — CHANGES_REQUESTED** [review]
Please fix lint issue

File: website/templates/grant_access.html [hunk 1]
```diff
@@ -0,0 +1,37 @@
+{% extends "base.html" %}
+{% load gravatar %}
+{% load static %}
+{% load humanize %}
+{% load custom_tags %}
```
Comment (arkid15r) [inline website/templates/grant_access.html:5]: Could you remove unused tags?

File: website/decorator.py [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+def private_test_function(user, user_pk, issue_pk):
+    try:
+        issue = Issue.objects.get(id=issue_pk)
+        if user.pk == user_pk:
+            if issue.user.pk == user.pk:
+                return True
+        return False
+
+    except:
```
Comment (arkid15r) [inline website/decorator.py:17]: This is not a good practice. Could you narrow down the except cases?

File: website/decorator.py [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        if user.pk == user_pk:
+            return True
+        return False
```
Comment (arkid15r) [inline website/decorator.py:20]: It looks like you can just `return user.pk == user_pk`

File: website/views.py [hunk 4]
```diff
@@ -341,6 +344,138 @@ def facebook_callback(request):
...
+    context = {"requesters": request_access, "viewers": issue.viewer.all()}
+
+    return render(request, "grant_access.html", context)
+
+
+@login_required(login_url="/accounts/login")
+def request_access(request, issue_pk):
+    try:
+        issue = Issue.objects.get(id=issue_pk)
+    except:
```
Comment (arkid15r) [inline website/views.py:436]: Should this should expect `Issue.DoesNotExist` only?

File: website/views.py [hunk 5]
```diff
@@ -341,6 +344,138 @@ def facebook_callback(request):
...
+        messages.error(request, "Issue Does not Exist")
+        return redirect("/")
+    try:
+        RequestIssueAccess.objects.get(user=request.user, issue=issue_pk)
+        messages.success(request, "Request has ALready been sent Please Wait for approval")
+        return redirect("/")
+    except:
```
Comment (arkid15r) [inline website/views.py:443]: Here too, please don't use broad excpeptions

**Review by arkid15r — COMMENTED** [review]
Please look into some code improvement suggestions:

**Comment by HanilJain** [comment]
> Please look into some code improvement suggestions:

@arkid15r Thanks for the suggestions they were really helpful.
if there is anything please tell.

File: website/migrations/0086_rename_requestaccess_requestissueaccess.py [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+class Migration(migrations.Migration):
+    dependencies = [
+        ("website", "0085_merge_20240218_1322"),
+        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
+    ]
+
+    operations = [
+        migrations.RenameModel(
+            old_name="RequestAccess",
+            new_name="RequestIssueAccess",
```
Comment (DonnieBLT) [inline website/migrations/0086_rename_requestaccess_requestissueaccess.py:16]: can you flatten these migrations so we don't require this rename?

File: website/models.py [hunk 7]
```diff
@@ -251,6 +251,7 @@ class Issue(models.Model):
     modified = models.DateTimeField(auto_now=True)
     is_hidden = models.BooleanField(default=False)
     rewarded = models.PositiveIntegerField(default=0)  # money rewarded by the company
+    viewer = models.ManyToManyField(User, null=True, blank=True, related_name="viewer")
```
Comment (DonnieBLT) [inline website/models.py:254]: don't we already have a viewer?

**Comment by Sarthak5598** [comment]
Great work @HanilJain , Happy to have someone this skilled as a peer

**Comment by HanilJain** [comment]
> Great work @HanilJain , Happy to have someone this skilled as a peer

Thanks @Sarthak5598 , but its nothing like that : )

**Review by DonnieBLT — CHANGES_REQUESTED** [review]
Please see a couple tweaks and concerns

File: website/migrations/0084_remove_issue_viewer_issue_viewer.py [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+
+
+class Migration(migrations.Migration):
+    dependencies = [
+        migrations.swappable_dependency(settings.AUTH_USER_MODEL),
+        ("website", "0083_issue_viewer"),
+    ]
+
+    operations = [
+        migrations.RemoveField(
```
[the remark this answers is not in the document]
Comment (HanilJain) [reply website/migrations/0084_remove_issue_viewer_issue_viewer.py:14]: Done 👍

File: website/models.py (hunk 7 above)
Comment (HanilJain) [reply website/models.py:255]: @DonnieBLT yup I have replied on it too.
viewer is field to give view access on is_hidden issues by the owner of issue.
there was no other field like this

File: website/models.py [hunk 9]
```diff
@@ -315,6 +316,12 @@ class Meta:
         ordering = ["-created"]
 
 
+class RequestIssueAccess(models.Model):
```
[the remark this answers is not in the document]
Comment (HanilJain) [reply website/models.py:319]: naming was suggested here https://github.com/OWASP-BLT/BLT/pull/1737#issuecomment-1957882402

File: website/templates/email/granted_access.txt [hunk 10]
```diff
@@ -0,0 +1,61 @@
+{% load custom_tags %}
```
[the remark this answers is not in the document]
Comment (HanilJain) [reply website/templates/email/granted_access.txt:1]: @DonnieBLT It mail content which will sent after an access is granted

**Review by HanilJain — COMMENTED** [review]
@DonnieBLT I have realised on the previous suggested changes i have replied to bu forget to submit review, hence not got posted

**Comment by HanilJain** [comment]
@DonnieBLT Made all suggested changes 7e2081ec067c42cd88735e9802a5b70a1c5ccffd

**Review by DonnieBLT — CHANGES_REQUESTED** [review]
Can you please split these into smaller PRs

**Comment by HanilJain** [comment]
> Can you please split these into smaller PRs

Yup i'll do it

**Comment by HanilJain** [comment]
@DonnieBLT  For sure I have planned do it in upcoming few days.

**Pull request closed without merging**